templates/CMakeLists.txt.template

```diff
@@ -486,8 +507,42 @@
 
   if(WIN32)
     set(_gRPC_ALLTARGETS_LIBRARIES <%text>${_gRPC_ALLTARGETS_LIBRARIES}</%text> ws2_32 crypt32)
+    set(_gRPC_STATIC_WIN32 STATIC)
   endif()
 
+  if(BUILD_SHARED_LIBS AND WIN32)
+
+  # Currently for shared lib on Windows (i.e. a DLL) certain bits of source code
+  # are generated from protobuf definitions by upbc. This source code does not include
+  # annotations needed to export these functions from gprc.lib so we have to
+  # re-include a small subset of these.
+  #
+  # This is not an ideal situation because these functions will be unavailable
+  # to clients of grpc and the libraries that need this (e.g. grpc++) will
+  # include redundant duplicate code. Hence, the duplication is only activated
+  # for DLL builds - and should be completely removed when source files are
+  # generated with the necessary __declspec annotations.
+  set(gRPC_UPB_GEN_DUPL_SRC
+    src/core/ext/upb-generated/src/proto/grpc/gcp/altscontext.upb.c
+    src/core/ext/upb-generated/src/proto/grpc/health/v1/health.upb.c
+    src/core/ext/upb-generated/src/proto/grpc/gcp/transport_security_common.upb.c
+  )
+
+  # When building shared lib on Windows, additional missing symbols need to
+  # be resolved for grpc++_unsecure. Also that build needs to be able to resolve
```

Comment:
Would you elaborate this problem? I think something is wrong if unsecure has to be linked with openssl.

Reply:
Even though @dawidcha has improved the implementation, I believe this goes back to something I did originally. As I tried to switch our arrow builds in conda-forge to depend on the shared builds of grpc, there was an experimental symbol (`grpc::experimental::StaticDataCertificateProvider`) that could not get found anymore.

When [trying](https://github.com/conda-forge/grpc-cpp-feedstock/pull/303) to add `GRPCXX_DLL` to that class, the build failed because it was relying on symbols that were only present in `grpc++` (rather than `grpc++_unsecure`). I didn't analyse this further as it didn't make a big difference on our side, but I guess the `StaticDataCertificateProvider` should be fully in `grpc++`?

Reply:
The reason this is happening is that there is a dependency on StaticDataCertificateProvider which is ultimately resolved by including the file src\core\lib\security\credentials\tls\grpc_tls_certificate_provider.cc. This source file includes a function PrivateKeyAndCertificateMatch which has a dependency on openssl, hence the need to link it to resolve. For static libraries, I assume the PrivateKeyAndCertificateMatch function is being excluded by the linker because it isn't being used - the DLL can't do that because as a runtime artifact the linker can't know whether or not the function will be used, so has to include it.

I believe that it should be possible to get round this by moving the PrivateKeyAndCertificateMatch function (and maybe a few others) into a separate compilation unit that isn't included in the xx_unsecure libraries.

I can do that now, or in a follow up PR.